test/cluster: Adjust tests to RF-rack-valid keyspaces

dawmd opened the pull request:
In this PR, we're adjusting most of the cluster tests so that they pass
with the `rf_rack_valid_keyspaces` configuration option enabled. In most
cases, the changes are straightforward and require little to no additional
insight into what the tests are doing or verifying. In some, however, doing
that does require a deeper understanding of the tests we're modifying.
The justification for those changes and their correctness is included in
the commit messages corresponding to them.

Note that this PR does not cover all of the cluster tests. There are few
remaining ones, but they require a bit more effort, so we delegate that
work to a separate PR.

I tested all of the modified tests locally with `rf_rack_valid_keyspaces`
set to true, and they all passed.

Fixes scylladb/scylladb#23959

Backport: we want to backport these changes to 2025.1 since that's the version where we introduced RF-rack-valid keyspaces in. Although the tests are not, by default, run with `rf_rack_valid_keyspaces` enabled yet, that will most likely change in the near future and we'll also want to backport those changes too. The reason for this is that we want to verify that Scylla works correctly even with that constraint.

dawmd commented:
Since the PR modifies a lot of tests and each of them may take some time to understand, I'd like to ask specific people to review the last three commits:

* @xtrey, could you review commit 3?
* @patjed41, could you review commits 4 and 5?

I'd really appreciate it. Commit 2 is extremely straightforward, so it can be reviewed by anyone. The same goes for commit 1—this one is quite big, though.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16670/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16670/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16670/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:small_blue_diamond: cluster/test_tablets_cql
:small_blue_diamond: cluster/test_tablets_migration
:small_blue_diamond: cluster/test_tls
#### Failed Tests (40/644):
- [test_global_ignored_nodes_list.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.1)
- [test_global_ignored_nodes_list.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_1_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.1)
- [test_global_ignored_nodes_list.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.10)
- [test_global_ignored_nodes_list.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_10_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.10)
- [test_global_ignored_nodes_list.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.2)
- [test_global_ignored_nodes_list.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_2_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.2)
- [test_global_ignored_nodes_list.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.3)
- [test_global_ignored_nodes_list.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_3_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.3)
- [test_global_ignored_nodes_list.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.4)
- [test_global_ignored_nodes_list.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16670/testReport/junit/cluster/test_global_ignore_nodes/test_global_ignored_nodes_list_debug_4_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_global_ignored_nodes_list.debug.4)
#### Build Details:
- Duration: 6 hr 55 min
- Builder: spider9.cloudius-systems.com

patjed41 commented on test/cluster/test_global_ignore_nodes.py:
```diff
@@ -21,8 +21,8 @@ async def test_global_ignored_nodes_list(manager: ManagerClient, random_tables)
        have two live nodes for the quorum.
     """
-    await manager.servers_add(2)
     servers = await manager.running_servers()
+    await manager.servers_add(2, property_file=[servers[1].property_file(), servers[2].property_file()])
```
Moving `servers_add` below `servers = await manager.running_servers()` is incorrect because `servers` won't contain the two new servers anymore. CI probably failed because of this.

patjed41 commented on test/cluster/test_read_repair.py:
```diff
@@ -382,11 +382,9 @@ async def test_read_repair_with_trace_logging(request, manager):
 
-    for i in range(2):
-        await manager.server_add(cmdline=cmdline, config=config)
+    srvs = await manager.server_add(cmdline=cmdline, config=config, auto_rack_dc="dc1")
```
Here you replace 2 servers with 1, another cause of a CI failure.

patjed41 commented on test/cluster/test_replace.py:
```diff
@@ -90,7 +90,7 @@ async def test_replace_reuse_ip(request, manager: ManagerClient) -> None:
     await manager.server_stop_gracefully(servers[0].server_id)
     replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = True, use_host_id = False)
-    replace_future = asyncio.create_task(manager.server_add(replace_cfg))
+    replace_future = asyncio.create_task(manager.server_add(replace_cfg, property_file=servers[0].property_file()))
```
It's a bit annoying that we must pass the property file now, even though the replacing node must be in the same DC and rack as the node being replaced (right?). I think we could do the following: if `replace_cfg` is passed and `property_file` is not passed to `server_add`, set `property_file` to the property file of the node being replaced (doable because we have access to `replace_cfg.replaced_id`). Of course, it's an optional change.

dawmd replied:
> even though the replacing node must be in the same DC and rack as the node being replaced (right?)

Yes. Otherwise, starting a node will fail with the following error:

```
ERROR 2025-04-11 18:46:41,047 [shard 0:main] init - Startup failed: std::runtime_error (Cannot replace node 583dbb91-47fc-4f98-a40d-808bfaf03e53/127.75.128.6 with a node on a different data center or rack. Current location=datacenter1/rack1, new location=DEFAULT_DC/DEFAULT_RACK)
```

It's a good idea to enforce using the DC/rack of the replaced node by default. However, it's proved to be a more subtle and annoying thing than I thought it would be, so let's do it in another PR. I created an issue for it: scylladb/scylladb#23711.

patjed41 reviewed (changes requested):
I rewieved the whole PR.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16768/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16768/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16768/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_data_resurrection_in_memtable
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_raft_recovery_user_data
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_basic
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16768/)
#### Failed Tests (3/42770):
- [test_random_failures[stop_after_setting_mode_to_normal_raft_topology-add_new_udt].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16768/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_setting_mode_to_normal_raft_topology_add_new_udt__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_setting_mode_to_normal_raft_topology-add_new_udt].debug.3)
- [test_random_failures[stop_after_starting_auth_service-add_new_keyspace].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16768/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_starting_auth_service_add_new_keyspace__debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_auth_service-add_new_keyspace].debug.1)
- [test_random_failures[stop_after_starting_commitlog-stop_non_coordinator_node_gracefully].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16768/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_starting_commitlog_stop_non_coordinator_node_gracefully__debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_commitlog-stop_non_coordinator_node_gracefully].debug.2)
#### Build Details:
- Duration: 15 hr
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16799/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16799/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16799/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/random_failures/test_random_failures
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_data_resurrection_in_memtable
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_raft_recovery_user_data
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
#### Failed Tests (4/6262):
- [test_random_failures[stop_after_saving_tokens-decommission_node].debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16799/testReport/junit/cluster.random_failures/test_random_failures/test_random_failures_stop_after_saving_tokens_decommission_node__debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_saving_tokens-decommission_node].debug.4)
- [test_random_failures[stop_after_saving_tokens-decommission_node].debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16799/testReport/junit/cluster.random_failures/test_random_failures/test_random_failures_stop_after_saving_tokens_decommission_node__debug_4_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_saving_tokens-decommission_node].debug.4)
- [test_random_failures[stop_after_starting_gossiper-stop_coordinator_node_gracefully].debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16799/testReport/junit/cluster.random_failures/test_random_failures/test_random_failures_stop_after_starting_gossiper_stop_coordinator_node_gracefully__debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_gossiper-stop_coordinator_node_gracefully].debug.10)
- [test_random_failures[stop_after_starting_gossiper-stop_coordinator_node_gracefully].debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16799/testReport/junit/cluster.random_failures/test_random_failures/test_random_failures_stop_after_starting_gossiper_stop_coordinator_node_gracefully__debug_10_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures[stop_after_starting_gossiper-stop_coordinator_node_gracefully].debug.10)
#### Build Details:
- Duration: 9 hr 12 min
- Builder: spider5.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16837/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16837/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16837/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/random_failures/test_random_failures
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_data_resurrection_in_memtable
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_raft_recovery_user_data
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
#### Failed Tests (38/5034):
- [test_truncate_during_topology_change.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.1)
- [test_truncate_during_topology_change.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.10)
- [test_truncate_during_topology_change.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_10_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.10)
- [test_truncate_during_topology_change.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.2)
- [test_truncate_during_topology_change.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_2_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.2)
- [test_truncate_during_topology_change.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.3)
- [test_truncate_during_topology_change.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_3_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.3)
- [test_truncate_during_topology_change.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.4)
- [test_truncate_during_topology_change.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_4_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.4)
- [test_truncate_during_topology_change.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16837/testReport/junit/cluster/test_tablets2/test_truncate_during_topology_change_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_truncate_during_topology_change.debug.5)
#### Build Details:
- Duration: 8 hr 50 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16849/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16849/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16849/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/random_failures/test_random_failures
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_data_resurrection_in_memtable
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_raft_recovery_user_data
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16849/)
#### Build Details:
- Duration: 15 hr
- Builder: spider1.cloudius-systems.com

patjed41 commented on test/pylib/repair.py:
```diff
@@ -51,7 +51,8 @@ async def create_table_insert_data_for_repair(manager, rf = 3 , tablets = 8, fas
     if disable_flush_cache_time:
         config.update({'repair_hints_batchlog_flush_cache_time_in_ms': 0})
-    servers = [await manager.server_add(config=config, cmdline=cmdline), await manager.server_add(config=config, cmdline=cmdline), await manager.server_add(config=config, cmdline=cmdline)]
+    servers = await manager.servers_add(3, config=config, cmdline=cmdline,
+                                        property_file=[{"dc": "dc1", "rack": f"r{i % rf}"} for i in range(rf)])
```
> If RF > 3, exactly three racks are created.

I don't understand why it's correct. Will creating a keyspace below succeed if `rf > 3` even though we create only 3 racks here? Maybe it won't but tests never do it. In that case, we should probably assert `rf <= 3` here.

dawmd replied:
Fair enough. I intentionally omitted the check to allow for running the function with RF > 3 in case `rf_rack_valid_keyspaces` is disabled (because someone needs to test that scenario for whatever reason), but you're right. There should be no point in that. I'll change it.

patjed41 commented on test/cluster/test_tablets.py:
```diff
@@ -137,7 +137,7 @@ async def test_reshape_with_tablets(manager: ManagerClient):
 @pytest.mark.asyncio
 async def test_tablet_rf_change(manager: ManagerClient, direction):
```
I guess the problem with this test was that with 3 racks RF=2 is not RF-rack-valid. You could add it to the commit message to make it clear why the changes are needed.

This test made me realize that we have the following problem: if we have n racks in a DC, where n > 2, and a keyspace with RF=0 or RF=1 in this DC, we cannot alter its RF to n because ALTER with tablets can change RF by only 1 and RF=2, 3, ..., n - 1 are not RF-rack-valid. Similarly, we cannot alter RF=n to RF=0 or RF=1. Are you aware of this problem? Maybe we should open an issue.

dawmd replied:
Yes, it's a known issue and is worked on, cf. scylladb/scylladb#17572 and scylladb/scylladb#17579.

patjed41 commented on test/cluster/random_failures/cluster_events.py:
```diff
@@ -474,7 +474,7 @@ async def add_new_node(manager: ManagerClient,
     yield
 
     LOGGER.info("Add a new node to the cluster")
```
I'm not sure how we should proceed with tests from this commit. We could open an umbrella issue containing a separate issue for each test. The problem is that these tests will probably be fixed by different people, so the umbrella issue would be without a real assignee (maybe it's normal, I don't know).

dawmd replied:
I'll create the issues once this PR is merged (so there's a context for them).

patjed41 reviewed (changes requested):
Minor comments and questions.

Next time please try to explain CI failures and provide a comment with changes in the new version before you request review. This time I easily managed without it, but it's not always the case.

dawmd commented:
Thank you, Patryk, for the review. For tracking reasons, I'm listing the changelong for each consecutive version of this PR:

Version 2:
* Minor technical fixes: `property_file` is a regular function of `ServerInfo`, not marked with `@property`; providing the property file in places where it was necessary and which I missed in version 1.

Version 3:
* Rebased on the contemporary master branch (including adjustments to RF-rack-validity in MV subsuite of the cluster tests).
* Minor technical fixes.
* Explicitly disabled `rf_rack_valid_keyspaces` in problematic tests.
* Adjusted missed tests detected by CI.

Version 4:
* Rebased on the contemporary master branch.
* Tried adjusting `test_random_failures`. I had some issues running it locally, and it seems my changes weren't sufficient since CI failed.

Version 5:
* Disabled `rf_rack_valid_keyspaces` in `test_random_failures`.

Version 6:
* Rebased on the contemporary master branch.
* Adjusted a new cluster test added after posting version 5 to RF-rack-validity.

dawmd commented:
Version 7:
* Added an assert in `test/pylib/repair.py`.
* Rebased on top of the current master branch.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16918/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16918/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16918/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/random_failures/test_random_failures
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_data_resurrection_in_memtable
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_raft_recovery_user_data
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16918/)
#### Build Details:
- Duration: 16 hr
- Builder: i-0be8a1076a7ccf9ae (m5d.16xlarge)

patjed41 commented:
The last CI run failed on #22501, I restarted CI

dawmd commented on test/pylib/repair.py:
```diff
@@ -45,13 +45,16 @@ async def load_tablet_repair_task_infos(cql, host, table_id):
     return repair_task_infos
 
 async def create_table_insert_data_for_repair(manager, rf = 3 , tablets = 8, fast_stats_refresh = True, nr_keys = 256, disable_flush_cache_time = False, cmdline = None) -> (list[ServerInfo], CassandraSession, list[Host], str, str):
+    assert rf <= 3, "A keyspace with RF > 3 will be RF-rack-invalid if there are fewer racks than 3"
```
Right, it was supposed to be RF

dawmd commented:
Version 8:
* Rebased on top of the current master branch.
* Corrected the comment.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16996/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16996/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16996/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/random_failures/test_random_failures
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_data_resurrection_in_memtable
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_raft_recovery_user_data
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16996/)
#### Build Details:
- Duration: 15 hr
- Builder: spider7.cloudius-systems.com

patjed41 reviewed (approved):
I'll queue it once the merge freeze ends

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17186/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17186/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17186/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/random_failures/test_random_failures
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_data_resurrection_in_memtable
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_raft_recovery_user_data
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5134/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8303/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16385/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17186/)
#### Build Details:
- Duration: 15 hr
- Builder: spider7.cloudius-systems.com

dawmd commented:
Version 9.1:
* Rebased on top of the current master.

dawmd commented:
Version 9.2:
* Marked `cluster/test_long_query_timeout_erm.py` as a test disabling `rf_rack_valid_keyspaces`. Adjusting it may be quite hard (at least it's not clear right off the bat how to do it), so let's do it in a separate PR.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17197/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17197/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17197/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cluster/object_store/test_backup
:small_blue_diamond: cluster/random_failures/test_random_failures
:small_blue_diamond: cluster/test_change_ip
:small_blue_diamond: cluster/test_change_rpc_address
:small_blue_diamond: cluster/test_conflicting_keys_read_repair
:small_blue_diamond: cluster/test_data_resurrection_in_memtable
:small_blue_diamond: cluster/test_fencing
:small_blue_diamond: cluster/test_global_ignore_nodes
:small_blue_diamond: cluster/test_long_query_timeout_erm
:small_blue_diamond: cluster/test_multidc
:small_blue_diamond: cluster/test_not_enough_token_owners
:small_blue_diamond: cluster/test_raft_recovery_user_data
:small_blue_diamond: cluster/test_read_repair
:small_blue_diamond: cluster/test_remove_rpc_client_with_pending_requests
:small_blue_diamond: cluster/test_repair
:small_blue_diamond: cluster/test_replace
:small_blue_diamond: cluster/test_reversed_queries_during_simulated_upgrade_process
:small_blue_diamond: cluster/test_rpc_compression
:small_blue_diamond: cluster/test_select_from_mutation_fragments
:small_blue_diamond: cluster/test_snapshot
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17197/)
#### Build Details:
- Duration: 15 hr
- Builder: spider1.cloudius-systems.com

patjed41 commented:
queued
